mirror/github/clone_repos.py

```diff
@@ -1,120 +1,282 @@
-@click.option('--token', '-t', help='Access token for increase rate limit. Read from $GITHUB_TOKEN if specify.', default='', show_default=True)
-@click.option('--amount', '-n', help='Amount of repo.', type=int, default=50, show_default=True)
-@click.option('--languages-file', '-f', help='Path to json file with languages for extracting.')
+                print("Rate limit is end. Awaiting 1 minute.")
+                time.sleep(60)
+            else:
+                break
+    return response
 
 
-def clone_repos(crawldir: str, stars_expression: str, languages: Tuple, token: str, amount: int, languages_file: str):
+def get_lang(repo):
     """
-    Clone repos from search api to output dir.
-    Be careful check of upload size not provide
+    Return name of output language folder
+    """
+    if isinstance(repo["language"], str):
+        return repo["language"]
+    elif isinstance(repo["language"], list):
+        if repo["language"]:
+            return repo["language"][0]
+    return "Without language"
+
+
+def check_command(name):
+    """
+    Check whether `name` is on PATH and marked as executable.
+    """
+
+    # from whichcraft import which
+    from shutil import which
+
+    return which(name) is not None
+
+
+def create_dir_meta_if_not_exists(lang_path: str, meta_file: str, lang: str):
+
+    """
+    Create nessesary structure
+    """
+
+    if not os.path.exists(lang_path):
+        os.makedirs(lang_path)
 
-    output structure:
-    - crawldir
-      - language 1
-        - repo 1
-        - repo 2
-        ...
-      - language 2
-        - repo 1
-        - repo 2
-        ...
-      ...
+    if not os.path.exists(meta_file):
+        with open(meta_file, "w") as meta:
+            json.dump(
+                {
+                    "language": lang,
+                    "repos": [],
+                    "crawled_at": None,
+                    "mirror version": module_version,
+                },
+                meta,
+            )
 
 
+def read_repos(repos_dir, file_name, start_id, end_id):
     """
+    Read repos from file. Filter repos by given repo id range if specified.
+    """
+    repos_file_path = os.path.join(repos_dir, file_name)
+
+    # load available repo
+    if os.path.isfile(repos_file_path):
+        with open(repos_file_path, "r") as repos_file:
+            if start_id and end_id:
+                return [
+                    repo for repo in json.loads(repos_file.read())["data"] if repo["id"]
+                ]
+            else:
+                return json.loads(repos_file.read())["data"]
+
+
+def get_repos_files(repos_dir, start_id, end_id):
+
+    """
+    Return list of files with repose by given ids range or all files from folder if ids range not set
+
+    In order to make sure that all repositories are covered,
+    add 2 additional files from the beginning of the ordered directory list and from the end
+
+    """
+
+    dir_files = os.listdir(repos_dir)
```

Comment:
Search creates a `languages_config.json` file which we should NOT be reading in get_repos_files.

Reply:
You can make an explicit rule which ignores that file as a quick fix.

Reply:
Actually, the issue with `languages_config.json` was that it had two JSON objects one after the other.